nixos/tandoor-recipes: Fix support for `SCRIPT_NAME` setting

l0b0 opened the pull request:
## Description of changes

- Run `tandoor-recipes` from within its `MEDIA_ROOT` directory.
- Verify that all `href` attributes emitted as part of the entrypoint
  page after logging in are reachable.

Closes #262857 - the goal of that issue can now be achieved by setting the following:

```nix
{
  services.tandoor-recipes.extraConfig = {
    SCRIPT_NAME = "/any/path";
    STATIC_URL = "${nodes.machine.services.tandoor-recipes.extraConfig.SCRIPT_NAME}/static/";
  };
}
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

l0b0 committed:
nixos/tandoor-recipes: Fix working directory

Run `tandoor-recipes` from within its `MEDIA_ROOT` directory to support
the `SCRIPT_NAME` setting.

Closes #262857.

Artturin commented:
@ofborg  test tandoor-recipes-script-name

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-serve-tandoor-recipes-from-a-sub-path/32451/2

ambroisie commented on nixos/tests/tandoor-recipes-script-name.nix:
```diff
@@ -0,0 +1,40 @@
+    # Get CSRF token for later requests
+    csrf_token = machine.succeed(f"grep csrftoken {cookie_jar_path} | cut --fields=7")
+
+    machine.succeed(f"DJANGO_SUPERUSER_PASSWORD='{password}' ${pkgs.tandoor-recipes}/lib/tandoor-recipes/manage.py createsuperuser --no-input --username='{username}' --email=nobody@example.com")
```
As explained in #287956, you should use `/var/lib/tandoor-recipes/tandoor-recipes-manage` (this might require `sudo` to switch to the `tandoor_recipes` user).

l0b0 replied:
That seems to be [undocumented](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%22tandoor-recipes-manage%22&type=code). It's also not the name used by the Tandoor manual, so there's no reason any user would realise that.

ambroisie replied:
1. It does just wrap `manage.py`.
2. It's the pattern that is used by other Django services in NixOS (e.g: `paperless`). Nextcloud also does a similar thing with its CLI administration tool.
3. The technical reason is to use the same environment variables (i.e: configuration) as the service. 
4. You can improve the documentation if you feel it is unclear.

l0b0 replied:
> (this might require `sudo` to switch to the `tandoor_recipes` user).

Does anything actually create the `tandoor_recipes` user? The main Nix file just [sets the service to run as that user](https://github.com/NixOS/nixpkgs/blob/2058a82fb0d09431e0c449bf29ea96c10c821da8/nixos/modules/services/misc/tandoor-recipes.nix#L87), but doesn't create it. And based on the [documentation](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#DynamicUser=), it looks like `DynamicUser = true;` means it just creates a random user at runtime.

l0b0 replied:
And `machine.succeed("/var/lib/tandoor-recipes/tandoor-recipes-manage migrate")` says "django.db.utils.OperationalError: unable to open database file".

ambroisie replied:
> Does anything actually create the tandoor_recipes user?

Systemd does, when the service is running.

I have recently [uploaded a change](#298168) (inspired by Photoprism's module) to attach to the service's namespace to make this work reliably. Something like `sudo /var/lib/tandoor-recipes/tandoor-recipes-manage` should work (you need the `sudo` to be able to read/execute the file IIRC).

ambroisie replied:
> And `machine.succeed("/var/lib/tandoor-recipes/tandoor-recipes-manage migrate")` says "django.db.utils.OperationalError: unable to open database file".

So I think your actual issue is not using `sudo /.../tandoor-recipes-manage`. I don't remember having any issues using the manage script on my server. Let me know if you have more issues. 

In trying it again locally, I noticed something.

I think I found an issue in my code, with the following message:

```
/var/lib/tandoor-recipes/tandoor-recipes-manage: line 16: UID: readonly variable
```

So it's not using `tandoor_recipes`' `UID`. I _think_ that's fine as it's probably just going to use `root`'s instead, which should have all the correct rights for DB access etc (as I said, I didn't have any issues when I used it)... But that's still not optimal.

l0b0 replied:
I can't find any reference to a `--wdns` argument; it doesn't seem to be available for the manage script at least:

```
usage: manage.py createsuperuser [-h] [--username USERNAME] [--noinput]
                                 [--database DATABASE] [--email EMAIL]
                                 [--version] [-v {0,1,2,3}]
                                 [--settings SETTINGS]
                                 [--pythonpath PYTHONPATH] [--traceback]
                                 [--no-color] [--force-color] [--skip-checks]
```

ambroisie replied:
It's a [`nsenter`](https://github.com/NixOS/nixpkgs/blob/e9ee548d90ff586a6471b4ae80ae9cfcbceb3420/nixos/modules/services/misc/tandoor-recipes.nix#L24) argument which is missing. I'm asking you to try adding it locally (to the module) and see if it fixes the SQLite issue.

l0b0 replied:
> It's a [`nsenter`](https://github.com/NixOS/nixpkgs/blob/e9ee548d90ff586a6471b4ae80ae9cfcbceb3420/nixos/modules/services/misc/tandoor-recipes.nix#L24) argument which is missing. I'm asking you to try adding it locally (to the module) and see if it fixes the SQLite issue.

Same issue:

```
❯ git diff
diff --git a/nixos/modules/services/misc/tandoor-recipes.nix b/nixos/modules/services/misc/tandoor-recipes.nix
index a2210f3d7db5..f3545fbbf530 100644
--- a/nixos/modules/services/misc/tandoor-recipes.nix
+++ b/nixos/modules/services/misc/tandoor-recipes.nix
@@ -22,7 +22,7 @@ let
     ${lib.toShellVars env}
     eval "$(${config.systemd.package}/bin/systemctl show -pUID,GID,MainPID tandoor-recipes.service)"
     exec ${pkgs.util-linux}/bin/nsenter \
-      -t $MainPID -m -S $UID -G $GID \
+      -t $MainPID -m -S $UID -G $GID --wdns="${pkg}" \
       ${pkg}/bin/tandoor-recipes "$@"
   '';
 in
diff --git a/nixos/tests/tandoor-recipes-script-name.nix b/nixos/tests/tandoor-recipes-script-name.nix
index f48695c38c53..5a14a219f9c2 100644
--- a/nixos/tests/tandoor-recipes-script-name.nix
+++ b/nixos/tests/tandoor-recipes-script-name.nix
@@ -22,6 +22,7 @@ import ./make-test-python.nix (
         inherit (nodes.machine.services.tandoor-recipes.extraConfig) SCRIPT_NAME;
       in
       ''
+        machine.shell_interact()
         username = "any_username"
         password = "any password"
```

```
$ cat /var/lib/tandoor-recipes/tandoor-recipes-manage
#!/nix/store/agkxax48k35wdmkhmmija2i2sxg8i7ny-bash-5.2p26/bin/bash
set -o allexport # Export the following env vars
DEBUG='0'
DEBUG_TOOLBAR='0'
GUNICORN_CMD_ARGS='--bind=localhost:8080'
MEDIA_ROOT='/var/lib/tandoor-recipes'
SCRIPT_NAME='/any/path'
STATIC_URL='/any/path/static/'
eval "$(/nix/store/cym39cl4v79bmbng4bvafkxagz96f4hw-systemd-255.6/bin/systemctl show -pUID,GID,MainPID tandoor-recipes.service)"
exec /nix/store/w8p87ilgm572kw000sbfqvfxziiw2sqz-util-linux-2.39.4-bin/bin/nsenter \
  -t $MainPID -m -S $UID -G $GID --wdns="/nix/store/gdv5fj6b33y8575wd9kfr9yaa0ng43x1-python3.11-tandoor-recipes-1.5.16" \
  /nix/store/gdv5fj6b33y8575wd9kfr9yaa0ng43x1-python3.11-tandoor-recipes-1.5.16/bin/tandoor-recipes "$@"
```

```
$ sudo DJANGO_SUPERUSER_PASSWORD='any password' /var/lib/tandoor-recipes/tandoor-recipes-manage createsuperuser --no-input --username='any_username' --email=nobody@example.com
machine # [  251.110671] sudo[975]:     root : TTY=hvc0 ; PWD=/tmp ; USER=root ; ENV=DJANGO_SUPERUSER_PASSWORD=any password ; COMMAND=/var/lib/tandoor-recipes/tandoor-recipes-manage createsuperuser --no-input --username=any_username --email=nobody@example.com
machine # [  251.156655] sudo[975]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=0)
machine # /var/lib/tandoor-recipes/tandoor-recipes-manage: line 10: UID: readonly variable
machine # Traceback (most recent call last):
machine #   File "/nix/store/7xvkzxrydfz6n0x18ik28h6fyn85f9lj-python3.11-django-4.2.12/lib/python3.11/site-packages/django/db/backends/base/base.py", line 289, in ensure_connection
machine #     self.connect()
machine #   File "/nix/store/7xvkzxrydfz6n0x18ik28h6fyn85f9lj-python3.11-django-4.2.12/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
machine #     return func(*args, **kwargs)
machine #            ^^^^^^^^^^^^^^^^^^^^^
machine #   File "/nix/store/7xvkzxrydfz6n0x18ik28h6fyn85f9lj-python3.11-django-4.2.12/lib/python3.11/site-packages/django/db/backends/base/base.py", line 270, in connect
machine #     self.connection = self.get_new_connection(conn_params)
machine #                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
machine #   File "/nix/store/7xvkzxrydfz6n0x18ik28h6fyn85f9lj-python3.11-django-4.2.12/lib/python3.11/site-packages/django/utils/asyncio.py", line 26, in inner
machine #     return func(*args, **kwargs)
machine #            ^^^^^^^^^^^^^^^^^^^^^
machine #   File "/nix/store/7xvkzxrydfz6n0x18ik28h6fyn85f9lj-python3.11-django-4.2.12/lib/python3.11/site-packages/django/db/backends/sqlite3/base.py", line 180, in get_new_connection
machine #     conn = Database.connect(**conn_params)
machine #            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
machine # sqlite3.OperationalError: unable to open database file
```

Trying to `cd /var/lib/tandoor-recipes` first produced the same result.

dotlambda reviewed (commented):
Please fix the commit message.

AndersonTorres commented on nixos/tests/all-tests.nix:
```diff
@@ -887,6 +887,7 @@ in {
   systemd-homed = handleTest ./systemd-homed.nix {};
   systemtap = handleTest ./systemtap.nix {};
   tandoor-recipes = handleTest ./tandoor-recipes.nix {};
+  tandoor-recipes-script-name = handleTest ./tandoor-recipes-script-name.nix {};
```
the commit message should be written as

```txt
nixosTests.tandoor-recipes-script-name: init
```

Besides I subscribe to ambroisie.

Artturin replied:
Done but doesn't build

l0b0 commented:
@ambroisie Welp, after actually testing what I created this PR for, it looks like the very first URL on the page is broken:

```
Verify that each <a> element's @href URL can be reached: ['http://localhost:8080', '/any/path/search/', '/any/path/plan/', '/any/path/shopping/', '/any/path/books/', '/any/path/accounts/login/', '/any/path/accounts/password/reset/']
machine: must succeed: curl --cookie /tmp/cookies.txt --cookie-jar /tmp/cookies.txt --fail --header 'Origin: http://localhost:8080' --show-error --silent http://localhost:8080
machine # [  201.564197] gunicorn[1391]: [2024-06-19 10:17:23 +0200] [1391] [ERROR] Error handling request /
machine # curl: (22) The requested URL returned error: 500
machine # [  201.568333] gunicorn[1391]: Traceback (most recent call last):
machine # [  201.568799] gunicorn[1391]:   File "/nix/store/7abksi7h0b8nwza1574fmfg0fmb4jcf0-python3.11-gunicorn-21.2.0/lib/python3.11/site-packages/gunicorn/workers/sync.py", line 135, in handle
machine # [  201.569534] gunicorn[1391]:     self.handle_request(listener, req, client, addr)
machine # [  201.569923] gunicorn[1391]:   File "/nix/store/7abksi7h0b8nwza1574fmfg0fmb4jcf0-python3.11-gunicorn-21.2.0/lib/python3.11/site-packages/gunicorn/workers/sync.py", line 168, in handle_request
machine # [  201.571401] gunicorn[1391]:     resp, environ = wsgi.create(req, client, addr,
machine # [  201.571816] gunicorn[1391]:                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
machine # [  201.573417] gunicorn[1391]:   File "/nix/store/7abksi7h0b8nwza1574fmfg0fmb4jcf0-python3.11-gunicorn-21.2.0/lib/python3.11/site-packages/gunicorn/http/wsgi.py", line 183, in create
machine # [  201.574003] gunicorn[1391]:     path_info = path_info.split(script_name, 1)[1]
machine # [  201.574464] gunicorn[1391]:                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
machine # [  201.574817] gunicorn[1391]: IndexError: list index out of range
```

ambroisie reviewed (changes requested):
LGTM, if the test passes I'm okay with this.

Could you please split the commit to have the following history:

1. `nixos/tandoor-recipes: fix working directory`
2. `nixosTests.tandoor-recipes-script-name: init`

ambroisie commented on nixos/tests/tandoor-recipes-script-name.nix:
```diff
@@ -0,0 +1,95 @@
+        username = "username"
+        password = "password"
+
+        print("Create admin user")
+        machine.succeed(
+            f"DJANGO_SUPERUSER_PASSWORD='{password}' /var/lib/tandoor-recipes/tandoor-recipes-manage createsuperuser --no-input --username='{username}' --email=nobody@example.com"
+        )
```
Ditto.

ambroisie commented on nixos/tests/tandoor-recipes-script-name.nix:
```diff
@@ -0,0 +1,95 @@
+            f"DJANGO_SUPERUSER_PASSWORD='{password}' /var/lib/tandoor-recipes/tandoor-recipes-manage createsuperuser --no-input --username='{username}' --email=nobody@example.com"
+        )
+
+        print("Get CSRF token for later requests")
+        csrf_token = machine.succeed(f"grep csrftoken {cookie_jar_path} | cut --fields=7").rstrip()
```
Ditto.

ambroisie commented on nixos/tests/tandoor-recipes-script-name.nix:
```diff
@@ -0,0 +1,95 @@
+        print("Get CSRF token for later requests")
+        csrf_token = machine.succeed(f"grep csrftoken {cookie_jar_path} | cut --fields=7").rstrip()
+
+        print("Log in as admin user")
+        machine.succeed(
+            f"{curl} --data 'csrfmiddlewaretoken={csrf_token}' --data 'username={username}' --data 'password={password}' {login_url}"
+        )
```
Ditto.

ambroisie commented on nixos/tests/tandoor-recipes-script-name.nix:
```diff
@@ -0,0 +1,95 @@
+            f"{curl} --data 'csrfmiddlewaretoken={csrf_token}' --data 'username={username}' --data 'password={password}' {login_url}"
+        )
+
+        print("Get the contents of the logged in main page")
+        logged_in_page = machine.succeed(f"{curl} --location {base_url}")
```
Ditto.

ambroisie commented on nixos/tests/tandoor-recipes-script-name.nix:
```diff
@@ -0,0 +1,95 @@
+        parser = UrlParser()
+        parser.feed(logged_in_page)
+
+        print(f"Verify that each <a> element's @href URL can be reached: {parser.urls}")
+        for url in parser.urls:
+            machine.succeed(f"{curl} {url}")
```
Ditto.

```suggestion
        for url in parser.urls:
            with subtest(f"Verify that {url} can be reached"):
                machine.succeed(f"{curl} {url}")
```

ambroisie commented on nixos/tests/tandoor-recipes-script-name.nix:
```diff
@@ -0,0 +1,95 @@
+        cookie_jar_path = "/tmp/cookies.txt"
+        curl = f"curl --cookie {cookie_jar_path} --cookie-jar {cookie_jar_path} --fail --header 'Origin: {origin_url}' --show-error --silent"
+
+        print("Wait for the service to respond")
+        machine.wait_for_unit("tandoor-recipes.service")
+        machine.wait_until_succeeds(f"{curl} {login_url}")
```
Use `subtest` please:

```suggestion
        with subtest("Wait for the service to respond"):
            machine.wait_for_unit("tandoor-recipes.service")
            machine.wait_until_succeeds(f"{curl} {login_url}")
```

l0b0 replied:
Subtests are only really relevant for the very last `machine.succeed(f"{curl} {url}")`. Everything else is just setup to verify that.

ambroisie reviewed (changes requested):
Actually, the test script can be improved.

l0b0 committed:
nixosTests.tandoor-recipes-script-name: init

Verify that all `href` attributes emitted as part of the entrypoint page
after logging in are reachable.

Co-authored-by: Bruno BELANYI <bruno@belanyi.fr>
